lib/events/pgevents/pgevents.go

```diff
@@ -0,0 +1,641 @@
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if usid == uuid.Nil {
+		return nil, trace.BadParameter("invalid all-zero session id")
+	}
+
+	var evs []events.EventFields
+	if err := l.beginTxFunc(ctx, txReadOnly, func(tx pgx.Tx) error {
+		evs = evs[:0]
+
+		var ev events.EventFields
+		_, err := tx.QueryFunc(ctx, `
+			SELECT event_data
+			FROM audit
+			WHERE session_id = $1 AND event_index >= $2
+			ORDER BY event_time, event_index, audit_id
+			LIMIT $3`,
+			[]any{usid, after, defaults.EventsMaxIterationLimit}, []any{&ev},
+			func(pgx.QueryFuncRow) error { evs = append(evs, ev); ev = nil; return nil },
+		)
+		return trace.Wrap(err)
+	}); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return evs, nil
+}
+
+// searchEvents returns events within the time range, filtering (optionally) by
+// event types, session id, and a generic condition, limiting results by a count
+// and by a maximum size of the underlying json data of
+// events.MaxEventBytesInResponse, sorting by time, session id and event index
+// either ascending or descending, and returning an opaque, URL-safe string that
+// can be passed to the same function to continue fetching data.
+func (l *Log) searchEvents(
+	ctx context.Context, fromUTC, toUTC time.Time,
+	eventTypes []string, cond *types.WhereExpr, sessionID string,
+	limit int, order types.EventOrder, startKey string,
+) ([]apievents.AuditEvent, string, error) {
+	if limit <= 0 {
+		limit = defaults.EventsIterationLimit
+	}
+	if limit > defaults.EventsMaxIterationLimit {
+		return nil, "", trace.BadParameter("limit %v exceeds %v", limit, defaults.MaxIterationLimit)
+	}
+
+	var startKeyset keyset
+	if startKey != "" {
+		if err := startKeyset.FromKey(startKey); err != nil {
+			return nil, "", trace.Wrap(err)
+		}
+	}
+
+	var condFn utils.FieldsCondition
+	if cond != nil {
+		f, err := utils.ToFieldsCondition(cond)
+		if err != nil {
+			return nil, "", trace.Wrap(err)
+		}
+		condFn = f
+	}
+
+	var q queryBuilder
+	q.Append(`
+		DECLARE cur CURSOR FOR SELECT
+			event_data,
+			event_time,
+			session_id,
+			event_index,
+			audit_id
+		FROM audit
+		WHERE event_time BETWEEN %v AND %v`,
+		fromUTC.UTC(), toUTC.UTC())
+	if len(eventTypes) > 0 {
+		q.Append(" AND event_type = ANY (%v)", eventTypes)
+	}
+	if len(sessionID) > 0 {
+		q.Append(" AND session_id = %v", sessionID)
+	}
+	if order != types.EventOrderDescending {
+		if len(startKey) > 0 {
+			q.Append(" AND (event_time, session_id, event_index, audit_id) > (%v, %v, %v, %v)",
+				startKeyset.t, startKeyset.sid, startKeyset.ei, startKeyset.id)
+		}
+		q.Append(" ORDER BY event_time, session_id, event_index, audit_id")
+	} else {
+		if len(startKey) > 0 {
+			q.Append(" AND (event_time, session_id, event_index, audit_id) < (%v, %v, %v, %v)",
+				startKeyset.t, startKeyset.sid, startKeyset.ei, startKeyset.id)
+		}
+		q.Append(" ORDER BY event_time DESC, session_id DESC, event_index DESC, audit_id DESC")
+	}
+	q.Append(" LIMIT %v", limit)
```

Comment:
Should we remove this? Should this be a static limit on `defaults.EventsMaxIterationLimit` (10k) instead, with the only limiting being done on our end after filtering out for the optional `cond`? I'm worried that the query might be optimized differently if we remove the limit - but then again, if we eventually implement server-side filtering, I don't see how an arbitrary query that looks through json can be optimized differently than a query that looks through everything with no limits.

Reply:
The way we do it in dynamoevents is to request by the limit but basically loop over and over again until we actually hit the correct limit post filtering. If we just remove the limit here we might load a humoungous amount of data which isn't too nice.

Reply:
That should be taken care of by the cursor tho, which will always fetch a chunk of a fixed size.

Reply:
Ah, in that case we're fine.

Reply:
Changed this to not have a fixed limit in the `SELECT`, and we just fetch small pages through the cursor until we hit the item limit, the total size limit, or the end of data.